modal/_partial_function.py

```diff
@@ -29,80 +30,163 @@
-    FUNCTION = 1
-    BUILD = 2
-    ENTER_PRE_SNAPSHOT = 4
-    ENTER_POST_SNAPSHOT = 8
-    EXIT = 16
-    BATCHED = 32
-    CLUSTERED = 64  # Experimental: Clustered functions
+    # Lifecycle method flags
+    BUILD = 1  # Deprecated, will be removed
+    ENTER_PRE_SNAPSHOT = 2
+    ENTER_POST_SNAPSHOT = 4
+    EXIT = 8
+    # Interface flags
+    CALLABLE_INTERFACE = 16
+    WEB_INTERFACE = 32
+    # Service decorator flags
+    # It's, unclear if we need these, as we can also generally infer based on some params being set
+    # In the current state where @modal.batched is used _insead_ of `@modal.method`, we need to give
+    # `@modal.batched` two roles (exposing the callable interface, adding batching semantics).
+    # But it's probably better to make `@modal.batched` and `@modal.method` stackable, or to move
+    # `@modal.batched` to be a class-level decorator since it primarily governs service behavior.
+    BATCHED = 64
+    CONCURRENT = 128
+    CLUSTERED = 256  # Experimental: Clustered functions
 
     @staticmethod
     def all() -> int:
         return ~_PartialFunctionFlags(0)
 
+    @staticmethod
+    def lifecycle_flags() -> int:
+        return (
+            _PartialFunctionFlags.BUILD  # Deprecated, will be removed
+            | _PartialFunctionFlags.ENTER_PRE_SNAPSHOT
+            | _PartialFunctionFlags.ENTER_POST_SNAPSHOT
+            | _PartialFunctionFlags.EXIT
+        )
+
+    @staticmethod
+    def interface_flags() -> int:
+        return _PartialFunctionFlags.CALLABLE_INTERFACE | _PartialFunctionFlags.WEB_INTERFACE
+
+
+@dataclass
+class _PartialFunctionParams:
+    webhook_config: Optional[api_pb2.WebhookConfig] = None
+    is_generator: Optional[bool] = None
+    force_build: Optional[bool] = None
+    batch_max_size: Optional[int] = None
+    batch_wait_ms: Optional[int] = None
+    cluster_size: Optional[int] = None
+    max_concurrent_inputs: Optional[int] = None
+    target_concurrent_inputs: Optional[int] = None
+    build_timeout: Optional[int] = None
+
+    def update(self, other: "_PartialFunctionParams") -> None:
+        """Update self with params set in other."""
+        for key, val in asdict(other).items():
+            if val is not None:
+                setattr(self, key, val)
+
 
 P = typing_extensions.ParamSpec("P")
 ReturnType = typing_extensions.TypeVar("ReturnType", covariant=True)
 OriginalReturnType = typing_extensions.TypeVar("OriginalReturnType", covariant=True)
+NullaryFuncOrMethod = Union[Callable[[], Any], Callable[[Any], Any]]
+NullaryMethod = Callable[[Any], Any]
 
 
 class _PartialFunction(typing.Generic[P, ReturnType, OriginalReturnType]):
-    """Intermediate function, produced by @enter, @build, @method, @web_endpoint, or @batched"""
+    """Object produced by a decorator in the `modal` namespace
+
+    The object will eventually by consumed by an App decorator.
+    """
 
-    raw_f: Callable[P, ReturnType]
+    obj: Callable[P, ReturnType]  # function, method, or class
```

Comment:
Wait, when do we allow `obj` to be a class? I think that would break a lot of assumptions in _PartialFunction and typing

Reply:
```
@app.cls()
@modal.concurrent()
class MyConcurrentService:
    ...
```